### ISSUE TITLE
FIR filters  44.1 48

### ISSUE THREAD

**Issue by bhack** [jaakkopasanen/AutoEq#356]
In the Readme we describe that we have FIR filters for 44.1 and 48 but it is not clear what we need to use.
Is the selection between 44.1 and 48 related to the source sampling frequency or the final output sampling of the audio server/soundcard?

**Comment by bhack** [comment]
It is similar to https://github.com/jaakkopasanen/AutoEq/issues/66 but I am asking this in the case we could improve the Easyeffect section of the Readme 

/cc @wmm

**Comment by wwmm** [comment]
The impulse response sampling rate has to match the sampling rate of the audio being convolved with the impulse. If they have different rates what is usually done is resampling the impulse response so it matches the incoming audio sampling rating.

**Comment by bhack** [comment]
So e.g. if we have a player at 44.1 and the  piperwire default at 48 if we load the impulse at 48 It will resampled at 44.1 and then to 48?

**Comment by wwmm** [comment]
>if we have a player at 44.1 and the piperwire default at 48 if we load the impulse at 48 It will resampled at 44.1 and then to 48?

No. If PipeWire is forced to work with 48 kHz What is going to happen is that the player output will be resampled to 48 kHz before its audio buffers get to the device it is linked to. EasyEffects filters and virtual devices follow the rate used by PipeWire. So they will be processing buffers at 48 kHz and the impulse won't be resampled because it is already at 48 kHz. 

But if you have dynamic sampling rate configured in PipeWire something different will happen. If the sound card is initially idle and the player plays something at 44.1 kHz PipeWire will put all filters and devices at 44.1 kHz. This time the player output won't be resampled. But the impulse response will have to be resampled to 44.1 kHz by the convolver plugin.

**Comment by bhack** [comment]
Thanks for the clarification, just an extra question.
As in the current Readme.md we have:

> Please keep in mind that not all EQ softwares support convolution. Some equalizers can load multiple FIR filters at the same time. Download both WAV files, create a Zip file containing both and load the Zip file to for example Roon.

Is the multiple FIR in convolution that you have recently added in Easyeffects master something related to this or It Is another feature? 
Does It make sense to automatically switch the native filter where available (with multiple loaded filters) instead of resampling?

**Comment by wwmm** [comment]
>Is the multiple FIR in convolution that you have recently added in Easyeffects master something related to this or It Is another feature?

In a way it is related. But not in the way you are thinking. The new convolver feature in EasyEffects allows the user to select two impulse response files and convolve them in a way that a new impulse response that has the characteristics of both impulses is generated and saved to a new file. The user still has to go through the usual process of selecting this new file as the impulse response to be applied to the signal by the convolver. 

In short this feature can not be used to select impulses based o their sampling rate. Honestly I would not be concerned about the impulse file resampling. When this is necessary it is an operation that happens once after the impulse is loaded and not every time a new audio buffer arrives.

**Comment by wwmm** [comment]
And I have my doubts that the impulse response resampling from 44.1 kHz to 48 kHz and vice versa will have noticeable effects over the quality of the output sound. It is probably going to be super hard or maybe even impossible to notice something like this by ear.

**Comment by bhack** [comment]
Thanks, I don't know why historically are distributed two "default" 44.1 kHz and 48 kHz filters here.

I only see this claim in the Readme.md

> ...The default results contain FIR filters for both 44.1 kHz and 48 kHz sampling rates. Other sampling rates are supported but not given in the default results.

**Comment by wwmm** [comment]
>I don't know why historically are distributed two "default" 44.1 kHz and 48 kHz filters here.

Probably because they are the most common. I see this as just a convenience for the users. I have my doubts that the people that create the impulse files redo the experiment with a different sampling rate. They probably measured with 48 kHz and resampled the output file to create the one with 44.1 kHz.

**Comment by bhack** [comment]
Generally convolutional filters (wav) could be generated with the tool in the repository from csv raw data measurements. The `-fs` arg are  frequencies requested, so you can produce multiple wave file yourself but the default distributed are 44.1 and 48. 

https://github.com/jaakkopasanen/AutoEq/blob/c352d1c1223b79f1daa7ab03b0a51cda86c33794/frequency_response.py#L774-L796

**Comment by jaakkopasanen** [comment]
Is there still an open question here?

AutoEq provides FIR filters with two different sampling rates for convenience since for example EqualizerAPO refuses to work with FIR filter WAV-file that has different sampling rate than the audio device. So in EqualizerAPO one should choose the one that matches whatever sampling rate has been configured for the audio device. Other systems might not be picky and there it probably doesn't matter.

Resampling of audio streams or FIR filters is not an audible issue.

**Comment by bhack** [comment]
Ok thanks
